lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,728 @@
+}
+
+// TODO: We should try to preserve type arguments on imported ObjC generic
+// classes, instead of relying on our knowledge of clang's encoding.
+// This would entail extracting the type arguments, calling `convert` to
+// create clang types, extracting the ObjCProtocolDecls and then using
+// getObjCObjectType with `id` as the base.
+clang::QualType
+ClangTypeConverter::visitBoundGenericClassType(BoundGenericClassType *type) {
+  // Any @objc class type in Swift that shows up in an @objc method maps 1-1 to
+  // "id <SomeProto>"; with clang's encoding ignoring the protocol list.
+  return getClangIdType(ClangASTContext);
+}
+
+clang::QualType
+ClangTypeConverter::visitBoundGenericType(BoundGenericType *type) {
+  // The only possibilities are *Pointer<T>, SIMD*<T>,
+  // ImplicitlyUnwrappedOptional<T> and Optional<T>.
+
+  if (!type->getDecl()->isObjC())
+    return clang::QualType();
+
+  if (type->getDecl()->getName().str() == "Optional") {
+    auto args = type->getGenericArgs();
+    if (args.size() != 1)
+      // We have a weird Optional with multiple generic arguments, eh?
+      return clang::QualType();
+    clang::QualType innerTy = convert(args[0]);
+    if (swift::canImportAsOptional(innerTy.getTypePtrOrNull()))
+      return innerTy;
+    return clang::QualType();
+  }
+
+  auto swiftStructDecl = type->getDecl();
+
+  enum class StructKind {
+    Invalid,
+    UnsafeMutablePointer,
+    UnsafePointer,
+    AutoreleasingUnsafeMutablePointer,
+    Unmanaged,
+    CFunctionPointer,
+    SIMD,
+  } kind = llvm::StringSwitch<StructKind>(swiftStructDecl->getName().str())
+    .Case("UnsafeMutablePointer", StructKind::UnsafeMutablePointer)
+    .Case("UnsafePointer", StructKind::UnsafePointer)
+    .Case("AutoreleasingUnsafeMutablePointer",
+          StructKind::AutoreleasingUnsafeMutablePointer)
+    .Case("Unmanaged", StructKind::Unmanaged)
+    .Case("CFunctionPointer", StructKind::CFunctionPointer)
+    .StartsWith("SIMD", StructKind::SIMD)
+    .Default(StructKind::Invalid);
+
+  auto args = type->getGenericArgs();
+  if (args.size() != 1)
+    // Must've got something other than *Pointer or SIMD*
+    return clang::QualType();
+  auto argCanonicalTy = args[0]->getCanonicalType();
+
+  switch (kind) {
+  case StructKind::Invalid:
+    return clang::QualType();
+
+  case StructKind::UnsafeMutablePointer:
+  case StructKind::Unmanaged:
+  case StructKind::AutoreleasingUnsafeMutablePointer: {
+    auto clangTy = convert(argCanonicalTy);
+    if (clangTy.isNull())
+      return clang::QualType();
+    return ClangASTContext.getPointerType(clangTy);
+  }
+  case StructKind::UnsafePointer: {
+    return ClangASTContext.getPointerType(convert(argCanonicalTy).withConst());
+  }
+
+  case StructKind::CFunctionPointer: {
+    auto &clangCtx = ClangASTContext;
+
+    clang::QualType functionTy;
+    if (isa<SILFunctionType>(argCanonicalTy->getCanonicalType())) {
+      functionTy = convert(argCanonicalTy);
```

Comment:
You actually assert on a `SILFunctionType` below.

Reply:
Yes, that is intentional -- if something fails to convert, its visit method should have an error, instead of substituting the convert+visit method at all potential call-sites which could trigger the assertion.